common-auth/src/main/java/com/hortonworks/registries/auth/server/KerberosAuthenticationHandler.java

```diff
@@ -417,4 +448,70 @@ public boolean shouldAuthenticate(HttpServletRequest request) {
         }
         return true;
     }
+
+    protected static String getDoasUser(HttpServletRequest request) {
+        String doAsUser = null;
+        String queryString = request.getQueryString();
+        if (queryString != null) {
+            String[] pairs = queryString.split(QUERY_STRING_DELIMITER);
+            try {
+                for (String pair : pairs) {
+                    if (pair.startsWith(DOAS_QUERY_STRING)) {
+                        doAsUser = URLDecoder.decode(pair.substring(DOAS_QUERY_STRING.length()), "UTF-8").trim();
+                        if (doAsUser.isEmpty()) {
+                            return null;
+                        }
+                    }
+                    break;
+                }
+            } catch (UnsupportedEncodingException ex) {
+                //We are providing "UTF-8". This should not be happening ideally.
+                LOG.error("Invalid encoding provided.");
+            }
+        }
+        return doAsUser;
+    }
+
+    /**
+     * Utility class to support proxy user authorizations.
+     *
+     * An example config will look like this:
+     * proxyuser.knox.hosts = 10.222.0.0
+     * proxyuser.admin.hosts = 10.222.0.0,10.113.221.221
+     */
+    class ProxyUserAuthorization {
+
+        /**
+         * Regex for the configuration property that indicates the Trusted Proxy user against approved host list.
+         */
+        static final String PROXYUSER_REGEX_PATTERN = "proxyuser\\.(.+)\\.hosts";
+        static final String DELIMITER = "\\s*,\\s*";
+
+        private Map<String, List<String>> proxyUserHostsMap = new HashMap<>();
+
+        ProxyUserAuthorization(Properties config) {
+            Pattern pattern = Pattern.compile(PROXYUSER_REGEX_PATTERN);
+            config.stringPropertyNames().forEach((propertyName -> {
+                Matcher matcher = pattern.matcher(propertyName);
+                if (matcher.find()) {
+                    String proxyUser = matcher.group(1);
+                    String hostList = config.getProperty(propertyName).trim();
+                    List<String> hosts = hostList.equals("*") ? new ArrayList<>() : Arrays.asList(hostList.split(DELIMITER));
+                    proxyUserHostsMap.put(proxyUser, hosts);
+                }
+            }));
+        }
+
+        public boolean authorize(String proxyUser, String host) {
+            List<String> proxyHosts = proxyUserHostsMap.get(proxyUser);
+            if (proxyHosts == null) {
+                return false;
+            } else if (proxyHosts.isEmpty()) {
+                return true;
+            }
+
+            return proxyHosts.contains(host);
+        }
```

Comment:
Can the user configure the upstream trusted proxy server as a regex? For example, **proxyuser.admin.hosts = 10.222.0.***